zarr/indexing.py

```diff
@@ -825,3 +826,116 @@ def pop_fields(selection):
+    return ls
+
+
+class PartialChunkIterator:
+    """Iterator to retrieve the specific coordinates of requested data
+    from within a compressed chunk.
+
+    Parameters
+    ----------
+    selection : tuple
+        tuple of slice objects to take from the chunk
+    arr_shape : shape of chunk to select data from
+
+    Attributes
+    ----------
+    arr_shape
+    selection
+
+    Returns
+    -------
+    Tuple with 3 elements:
+
+    start: int
+        ? bytes offset in the chunk to read from 
+    nitems: int
+        ? number of (bytes? elements?) to read in the chunk from start
+    partial_out_selection: list of slices
+        ? indices of the final selection to assign to after the partial read.
+
+    Notes
+    -----
+
+    My understanding of the partial read is that it create an empty array of the final shape of the required selection; 
+    and then do partial reading (with hacks in Blosc to do only partial decompress). 
+    And assign thos partial reads into the final partial buffer.
+    """
+
+    def __init__(self, selection, arr_shape):
+        ## TODO: self.selection seem to be unused outside of this
+        # and seem to be mutated; which might not be expected
+        # self.selection seem to be expected to be part of the external API of 
+        # this iterator; so not sure we can make it a local variable.
+        self.selection = make_slice_selection(selection)
+        self.arr_shape = arr_shape
+
+        # number of selection dimensions can't be greater than the number of chunk dimensions
+        if len(self.selection) > len(self.arr_shape):
+            raise ValueError(
+                "Selection has more dimensions then the array:\n"
+                "selection dimensions = {len(self.selection)\n"
+                "array dimensions = {len(self.arr_shape)}"
+            )
+
+        # any selection can not be out of the range of the chunk
+        selection_shape = np.empty(self.arr_shape)[tuple(self.selection)].shape
+        if any(
+            [
+                selection_dim < 0 or selection_dim > arr_dim
+                for selection_dim, arr_dim in zip(selection_shape, self.arr_shape)
+            ]
+        ):
+            raise IndexError(
+                "a selection index is out of range for the dimension"
+            )  # pragma: no cover
+
+        for i, dim_size in enumerate(self.arr_shape[::-1]):
+            index = len(self.arr_shape) - (i + 1)
+            if index <= len(self.selection) - 1:
+                slice_size = selection_shape[index]
+                if slice_size == dim_size and index > 0:
+                    self.selection.pop()
+                else:
+                    break
+
+        chunk_loc_slices = []
+        last_dim_slice = None if self.selection[-1].step > 1 else self.selection.pop()
+        for i, sl in enumerate(self.selection):
+            dim_chunk_loc_slices = []
+            for i, x in enumerate(slice_to_range(sl, arr_shape[i])):
```

Comment:
you use `i` in both loop and the argument of the second depends on I; that seem weird.

Reply:
inner i seem unused as well.